File: web/src/protocol/linkParams.ts
/// 配对链接解析 — TS 移植自 lib/protocol/link_params.dart。
/// 形如 https://zcode.z.ai/remote/v4?sid=...&hash=...&t=...&mid=...&name=...&app_version=...

export interface LinkParams {
  deviceSid: string
  passHash: string
  deviceMid: string | null
  deviceName: string | null
  appVersion: string | null
  source: URL
}

function get(uri: URL, key: string): string | null {
  const v = uri.searchParams.get(key)?.trim()
  return v ? v : null
}

export function parseLinkParams(raw: string): LinkParams | null {
  let uri: URL
  try {
    uri = new URL(raw.trim())
  } catch {
    return null
  }
  if (uri.protocol !== 'https:' && uri.protocol !== 'wss:') return null
  const sid = get(uri, 'sid')
  const hash = get(uri, 'hash')
  const t = Number(get(uri, 't') ?? '')
  if (!sid || !hash || !Number.isFinite(t) || t <= 0) return null
  return {
    deviceSid: sid,
    passHash: hash,
    deviceMid: get(uri, 'mid'),
    deviceName: get(uri, 'name'),
    appVersion: get(uri, 'app_version'),
    source: uri,
  }
}

/** relay websocket 地址：wss(s)://<host>/ws（带 mid 时挂 query）。 */
export function relayWsUri(p: LinkParams): string {
  const secure = p.source.protocol === 'https:' || p.source.protocol === 'wss:'
  const scheme = secure ? 'wss' : 'ws'
  const base = `${scheme}://${p.source.host}/ws`
  return p.deviceMid ? `${base}?mid=${encodeURIComponent(p.deviceMid)}` : base
}

File: lib/protocol/link_params.dart
/// Parses a ZCode web-remote connection URL, e.g.
/// https://zcode.z.ai/remote/v4?sid=...&hash=...&t=...&mid=...&name=...&app_version=...
class LinkParams {
  final String deviceSid;
  final String passHash;
  final String? deviceMid;
  final String? deviceName;
  final String? appVersion;
  final Uri source;

  const LinkParams({
    required this.deviceSid,
    required this.passHash,
    required this.source,
    this.deviceMid,
    this.deviceName,
    this.appVersion,
  });

  static String? _get(Uri uri, String key) {
    final v = uri.queryParameters[key]?.trim();
    return v == null || v.isEmpty ? null : v;
  }

  static LinkParams? parse(String raw) {
    Uri uri;
    try {
      uri = Uri.parse(raw.trim());
    } catch (_) {
      return null;
    }
    final sid = _get(uri, 'sid');
    final hash = _get(uri, 'hash');
    final t = int.tryParse(_get(uri, 't') ?? '');
    if ((uri.scheme != 'https' && uri.scheme != 'wss') ||
        sid == null ||
        hash == null ||
        t == null) {
      return null;
    }
    return LinkParams(
      deviceSid: sid,
      passHash: hash,
      deviceMid: _get(uri, 'mid'),
      deviceName: _get(uri, 'name'),
      appVersion: _get(uri, 'app_version'),
      source: uri,
    );
  }

  /// Relay websocket URL: `${ws(s)://<host>/ws` plus `?mid=` when present.
  Uri get relayWsUri {
    final scheme = secure ? 'wss' : 'ws';
    final base = Uri(
      scheme: scheme,
      host: source.host,
      port: source.hasPort ? source.port : null,
      path: '/ws',
    );
    if (deviceMid == null) return base;
    return base.replace(queryParameters: {'mid': deviceMid});
  }

  bool get secure => source.scheme == 'https' || source.scheme == 'wss';
}
